### PR TITLE
Support custom cost factors

### PR DESCRIPTION
**Pull request by taion**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -80,25 +79,33 @@ export class ComplexityVisitor {
...
 
   enterField() {
-    this.enterType(this.context.getType());
+    this.costFactor *= this.getFieldCostFactor();
+    this.getCalculator().addImmediate(this.costFactor * this.getFieldCost());
   }
 
-  enterType(type) {
+  leaveField() {
+    this.costFactor /= this.getFieldCostFactor();
```
Comment (jquense) [inline src/index.js]: every time, i forget this operator exists

File: test/fixtures/schema.js [hunk 2]
```diff
@@ -1,13 +1,26 @@
...
 } from 'graphql';
 
 const Item = new GraphQLObjectType({
   name: 'Item',
   fields: () => ({
     name: { type: GraphQLString },
     item: { type: Item },
+    expensiveItem: {
+      type: Item,
+      getCost: () => 50,
```
Comment (jquense) [inline test/fixtures/schema.js]: is this safe to extend like this?

Comment (taion) [reply test/fixtures/schema.js]: well, it works, doesn’t it?

**Comment by mxstbr** [comment]
How is this going to work if I write in the SDL?

**Comment by mxstbr** [comment]
> Does it make sense to set a cost factor in a non list type?

Yeah for sure, what if one of my fields has to do 10 db queries instead of the usual 1?

```GraphQL
type Bla {
  normalField: String # Cheap
  someExtremelyExpensiveField: SomeExtremelyExpensiveObject # Much more expensive
}
```

**Comment by itajaja** [comment]
wouldn't that be modeled by a constant cost on the field?

**Comment by taion** [comment]
I can't think of a use case for cost factor on a non-list field, but I don't think it really matters. I don't see the harm in keeping it around.

**Comment by taion** [comment]
@mxstbr I just added SDL support in the latest commit on this PR.

**Comment by taion** [comment]
Released in v0.2.2.

**Pull request merged**